matching brackets

anmolchabbra opened the pull request.

Commit:
matching brackets

Commit:
logic improved

Commit:
Add unit tests and convert to gradle.

Commit:
Indexerrorcorrected

bbhopesh commented on src/main/java/stack/MatchingBrackets.java:
```diff
@@ -0,0 +1,48 @@
+                    } else {
+                        return i + 1 ;
+                    }
+                    index = (string.indexOf(topElement));
```
There are multiple issues with doing `string.indexOf(topElement)`.
1. You are always calculating index when you see a closing bracket even when current closing bracket is balancing some open bracket. That is incorrect. Think about this example and see what your code returns on it `foo([]{}`  
2. If a character appears multiple times, `indexof` returns index of first occurence from the start of the string. You might not want this always. Think about this example `foo{}{`
3. `indexof`returns zero-based index while you want 1-based index.

bbhopesh commented on src/main/java/stack/MatchingBrackets.java:
```diff
@@ -0,0 +1,48 @@
+        if (s == -1) {
+            System.out.println("Success");
+        } else {
+            System.out.println(matcher.balanced(string));
```
Why are you calling `matcher.balanced` again here? Can't you use `s`?

bbhopesh commented on src/test/java/stack/MatchingBracketsTest.java:
```diff
@@ -0,0 +1,77 @@
+    public void negativeTestCase3() {
+        String string = "foo({{}[]}";
+        MatchingBrackets matcher = new MatchingBrackets();
+        assertEquals(10, matcher.balanced(string));
```
Why is expected value 10? Shouldn't it be 4?

Commit:
indexvaluecorrected

bbhopesh commented on src/main/java/stack/MatchingBrackets.java:
```diff
@@ -22,21 +22,20 @@ public int balanced(String string) {
             return -1;
         } else {
-           return index;
+            i = (string.lastIndexOf(stack.peek()) + 1);
```
This will still not work. Think about example `{{{}}`. What should your code returns and what should it return?

Don't always wait for me to give counter examples, whenever you make a change test it on a number of examples before you commit it. Did you test this change on multiple examples?

bbhopesh replied:
Here is a hint:
Because of duplicate characters, you cannot use `indexof` or `lastIndexof` as they will either give you first or last index of the character while answer could be first, last or any of the middle index depending on example. You will have to remember the index(es) in some data structure as you are processing the string.

Commit:
index are stored in stacks now

Commit:
using two stacks

Commit:
Following changes:

1. One class to group opening bracket and index.
2. Used new class to use only stack.
3. Move int i = 0 inside loop as we don't need i outside loop.
